### PR TITLE
Improve effective gas price implementation (RPC filter, gasprice/balance opcodes flags, new formula)

### LINKED ISSUE #2596: Improve effective gas price implementation (RPC filter, gasprice/balance opcodes flags, new formula)
# Implementation

Do changes in the effective gas price implementation to match the following (reviewed) process workflow:

https://app.diagrams.net/#G1rb3FKaEg4fnmu6mMdvvr-h36FjzEY4ek#%7B%22pageId%22%3A%22OfOgGhgHDn_CtmajkobP%22%7D

EGP second review-Workflow drawio

EGP second review-Formulas drawio

The main changes are:
- Implement a prefilter of the breackEvenGasPrice at RPC level
- Avoid the possible third execution of a tx reviewing the `has_gasprice_opcode` and `has_balance_opcode` returned by the executor after the first execution
- Use EffectiveGasPriceFinal formula to calculate the effectiveGasPrice when executing/comparing the tx at Sequencer level
- Store in the DB if the tx was executed a second time and which was the deviation with the initial EGPF
- To calculate the gas needed to store in L1 the tx data we must use 2 config values: `ByteGasCost` (16) for bytes that are not zero and `ZeroByteGasCost` (4) for bytes that are zero

Other minor changes:
- Rename const `totalRlpFieldsLength` to `constBytesTx`
- Rename config param `Margin` to `NetProfit`
- Add config param `BreakEvenFactor`

### PR DESCRIPTION
**Pull request by agnusmor**
Closes #2596 

### What does this PR do?

Do changes in the effective gas price implementation to match redefined process workflow. Check related issue for details

The following configuration parameters has been added:
> [Pool.EffectiveGasPrice]
> Enabled = true
> L1GasPriceFactor = 0.25
> ByteGasCost = 16
> ZeroByteGasCost = 4
> NetProfit = 1
> BreakEvenFactor = 1.1			
> FinalDeviationPct = 10

The following configuration parameters has beed deleted:
> [Sequencer.EffectiveGasPrice]
> MaxBreakEvenGasPriceDeviationPercentage = 10
> L1GasPriceFactor = 0.25
> ByteGasCost = 16
> MarginFactor = 1
> Enabled = false

### Reviewers

Main reviewers:

- @ToniRamirezM 
- @joanestebanr 
- @tclemos 
- @Psykepro

### CODE REVIEW

File: pool/pool.go [hunk 1]
```diff
@@ -221,12 +226,66 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+		} else {
+			// accept loss
+			loss = loss.Sub(breakEvenGasPriceWithFactor, tx.GasPrice())
+		}
+	}
+
+	log.Infof("egp-log: tx.GasPrice(): %v, breakEvenGasPrice: %v, breakEvenGasPriceWithFactor: %v, gasUsed: %v, reject: %t, loss: %v, L1GasPrice: %d, L2GasPrice: %d, Enabled: %t, tx: %s",
+		tx.GasPrice(), breakEvenGasPrice, breakEvenGasPriceWithFactor, preExecutionGasUsed, reject, loss, gasPrices.L1GasPrice, gasPrices.L2GasPrice, p.cfg.EffectiveGasPrice.Enabled, tx.Hash().String())
+
+	// Reject transaction if EffetiveGasPrice is enabled
```
Comment (ToniRamirezM) [inline pool/pool.go]:
```suggestion
	// Reject transaction if EffectiveGasPrice is enabled
```

File: proto/src/proto/executor/v1/executor.proto [hunk 2]
```diff
@@ -53,6 +54,8 @@ message ProcessBatchResponse {
     uint64 flush_id = 16;
     uint64 stored_flush_id = 17;
     string prover_id = 18;
+    repeated bytes nodes_keys = 19;
```
Comment (ToniRamirezM) [inline proto/src/proto/executor/v1/executor.proto:57]: ???

File: proto/src/proto/executor/v1/executor.proto [hunk 3]
```diff
@@ -53,6 +54,8 @@ message ProcessBatchResponse {
...
+    repeated bytes program_keys = 20;
```
Comment (ToniRamirezM) [inline proto/src/proto/executor/v1/executor.proto:58]: ???

File: proto/src/proto/executor/v1/executor.proto [hunk 4]
```diff
@@ -168,6 +171,8 @@ message Contract {
     string value = 3;
     bytes data = 4;
     uint64 gas = 5;
+    // Define type of internal call: CREATE, CREATE2, CALL, CALLCODE, DELEGATECALL, STATICCALL
+    string type = 6;
```
Comment (ToniRamirezM) [inline proto/src/proto/executor/v1/executor.proto:175]: ???

File: proto/src/proto/executor/v1/executor.proto [hunk 5]
```diff
@@ -473,4 +482,6 @@ enum ExecutorError {
     EXECUTOR_ERROR_INVALID_CONTRACTS_BYTECODE_KEY = 78;
     // EXECUTOR_ERROR_INVALID_CONTRACTS_BYTECODE_VALUE indicates that the input parameter contracts_bytecode value is invalid
     EXECUTOR_ERROR_INVALID_CONTRACTS_BYTECODE_VALUE = 79;
+    // EXECUTOR_ERROR_INVALID_GET_KEY indicates that the input parameter get key is invalid, e.g. is true but fork_id<5
+    EXECUTOR_ERROR_INVALID_GET_KEY = 80;
```
Comment (ToniRamirezM) [inline proto/src/proto/executor/v1/executor.proto:486]: The handling of this new error has not been added to /state/runtime/executor/errors.go

File: sequencer/finalizer.go [hunk 6]
```diff
@@ -571,46 +572,63 @@ func (f *finalizer) processTransaction(ctx context.Context, tx *TxTracker) (errW
 		f.processRequest.Transactions = tx.RawTx
 		hashStr = tx.HashStr
 
-		log.Infof("EffectiveGasPriceProcessCount=%d", tx.EffectiveGasPriceProcessCount)
-		// If it is the first time we process this tx then we calculate the BreakEvenGasPrice
-		if tx.EffectiveGasPriceProcessCount == 0 {
+		// If it is the first time we process this tx then we calculate the EffectiveGasPrice
+		if !reprocessTx {
```
Comment (ToniRamirezM) [inline sequencer/finalizer.go]: Maybe it is better to rename this variable to "firstTXProcess"

File: pool/effectivegasprice.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+// NewEffectiveGasPrice creates and initializes an instance of EffectiveGasPrice
+func NewEffectiveGasPrice(cfg EffectiveGasPriceCfg, minGasPriceAllowed uint64) *EffectiveGasPrice {
+	return &EffectiveGasPrice{
+		cfg:                cfg,
+		minGasPriceAllowed: minGasPriceAllowed,
+	}
+}
+
+// IsEffectiveGasPriceEnabled return if effectiveGasPrice calculation is enabled
+func (e *EffectiveGasPrice) IsEffectiveGasPriceEnabled() bool {
```
Comment (ToniRamirezM) [inline pool/effectivegasprice.go]:
```suggestion
func (e *EffectiveGasPrice) IsEnabled() bool {
```

File: sequencer/finalizer.go [hunk 8]
```diff
@@ -324,6 +320,7 @@ func (f *finalizer) addPendingTxToStore(ctx context.Context, txToStore transacti
 // finalizeBatches runs the endless loop for processing transactions finalizing batches.
 func (f *finalizer) finalizeBatches(ctx context.Context) {
 	log.Debug("finalizer init loop")
+	notFoundFirst := true // used to log debug only the first message when there is no txs to process
```
Comment (ToniRamirezM) [inline sequencer/finalizer.go]: The name of this variable does not transmit (at least to me) what it represents

File: state/types.go [hunk 9]
```diff
@@ -78,6 +78,36 @@ type ProcessTransactionResponse struct {
...
+	ValueFinal     *big.Int
+	ValueFirst     *big.Int
+	ValueSecond    *big.Int
+	FinalDeviation *big.Int
+	MaxDeviation   *big.Int
+	GasUsedFirst   uint64
+	GasUsedSecond  uint64
+	GasPrice       *big.Int
+	Percentage     uint8
+	Reprocess      bool
```
Comment (ToniRamirezM) [inline state/types.go:99]:
```suggestion
	NeedsReprocess      bool
```

**Review by ToniRamirezM — CHANGES_REQUESTED** [review]

File: sequencer/finalizer.go (hunk 6 above)
Comment (agnusmor) [reply sequencer/finalizer.go]: ok

File: pool/effectivegasprice.go (hunk 7 above)
Comment (agnusmor) [reply pool/effectivegasprice.go]: It's true!

File: state/types.go (hunk 9 above)
Comment (agnusmor) [reply state/types.go:99]: I undertand that using "re" prefix means that you do a second process, maybe "reprocessed"?

**Review by ToniRamirezM — APPROVED** [review]

**Pull request merged**